api/client/commands.go

```diff
@@ -2587,3 +2588,100 @@ func (cli *DockerCli) CmdExec(args ...string) error {
 
 	return nil
 }
+
+type containerStats struct {
+	Name             string
+	CpuPercentage    float64
+	Memory           float64
+	MemoryLimit      float64
+	MemoryPercentage float64
+	NetworkRx        float64
+	NetworkTx        float64
+}
+
+func (cli *DockerCli) CmdStats(args ...string) error {
+	cmd := cli.Subcmd("stats", "CONTAINER", "Stream the stats of a container", true)
+	cmd.Require(flag.Min, 1)
+	utils.ParseFlags(cmd, args, true)
+
+	cStats := map[string]containerStats{}
+	for _, name := range cmd.Args() {
+		if err := cli.streamStats(name, cStats); err != nil {
+			return err
+		}
+	}
+	w := tabwriter.NewWriter(cli.out, 20, 1, 3, ' ', 0)
+	for _ = range time.Tick(500 * time.Millisecond) {
+		fmt.Fprint(cli.out, "\033[2J")
+		fmt.Fprint(cli.out, "\033[H")
+		fmt.Fprintln(w, "CONTAINER\tCPU %\tMEM USAGE/LIMIT\tMEM %\tNET I/O")
+		for _, s := range sortStatsByName(cStats) {
+			fmt.Fprintf(w, "%s\t%.2f%%\t%s/%s\t%.2f%%\t%s/%s\n",
+				s.Name,
+				s.CpuPercentage,
+				units.HumanSize(s.Memory), units.HumanSize(s.MemoryLimit),
+				s.MemoryPercentage,
+				units.HumanSize(s.NetworkRx), units.HumanSize(s.NetworkTx))
+		}
+		w.Flush()
+	}
+	return nil
+}
+
+func (cli *DockerCli) streamStats(name string, data map[string]containerStats) error {
+	stream, _, err := cli.call("GET", "/containers/"+name+"/stats", nil, false)
+	if err != nil {
+		return err
+	}
+	go func() {
+		defer func() {
+			stream.Close()
+			delete(data, name)
+		}()
+
+		var (
+			previousCpu    uint64
+			previousSystem uint64
+			start          = true
+			dec            = json.NewDecoder(stream)
+		)
+		for {
+			var v *stats.Stats
+			if err := dec.Decode(&v); err != nil {
+				log.Error("decode container stat: %v", err)
+				return
+			}
+			var (
+				memPercent = float64(v.MemoryStats.Usage) / float64(v.MemoryStats.Limit) * 100.0
+				cpuPercent = 0.0
+			)
+			if !start {
+				cpuPercent = calcuateCpuPercent(previousCpu, previousSystem, v)
+			}
+			start = false
+			d := data[name]
```

Comment:
I'm not familiar with `map` concurrency guarantees, but is that a safe thing to do without locking?